scikeras/utils/transformers.py

```diff
@@ -0,0 +1,201 @@
+                # small sample sizes!
+                # don't even bother inverse transforming, just fill.
+                class_predictions = np.full(
+                    shape=(y.shape[0], 1), fill_value=self.classes_[0]
+                )
+            else:
+                if y.shape == 1 or (y.shape[1] == 1 and self.n_classes_ == 2):
+                    # result from a single sigmoid output
+                    # reformat so that we have 2 columns
+                    y = np.column_stack([1 - y, y])
+                y_ = np.argmax(y, axis=1).reshape(-1, 1)
+                class_predictions = self.final_encoder_.inverse_transform(y_)
+        elif self._target_type == "multiclass":
+            # array([0.8, 0.1, 0.1], [.1, .8, .1]) ->
+            # array(['apple', 'orange'])
+            idx = np.argmax(y, axis=-1)
+            if not is_categorical_crossentropy(self.loss):
+                y_ = idx.reshape(-1, 1)
+            else:
+                y_ = np.zeros(y.shape, dtype=int)
+                y_[np.arange(y.shape[0]), idx] = 1
+            class_predictions = self.final_encoder_.inverse_transform(y_)
+        elif self._target_type == "multiclass-onehot":
+            # array([.8, .1, .1], [.1, .8, .1]) ->
+            # array([[1, 0, 0], [0, 1, 0]])
+            idx = np.argmax(y, axis=-1)
+            y_ = np.zeros(y.shape, dtype=int)
+            y_[np.arange(y.shape[0]), idx] = 1
+            class_predictions = y_
+        else:
+            class_predictions = None
+
+        if return_proba:
+            return y
+        return np.squeeze(np.column_stack(class_predictions)).astype(
+            self.y_dtype_, copy=False
+        )
+
+    def get_metadata(self):
+        return {
+            "classes_": self.classes_,
+            "n_classes_": self.n_classes_,
+            "n_outputs_": self.n_outputs_,
+            "model_n_outputs_": self.model_n_outputs_,
+        }
+
+
+class RegressorTargetEncoder(BaseEstimator, TransformerMixin):
+    """Default target transformer for KerasRegressor.
+    """
+
+    def fit(self, X: np.ndarray) -> "RegressorTargetEncoder":
+        y = X  # rename for clarity, the input is always expected to be a target `y`
+        self.y_dtype_ = y.dtype
+        self.n_outputs_ = 1 if y.ndim == 1 else y.shape[1]
+        self.model_n_outputs_ = 1
+        return self
+
+    def transform(self, X: np.ndarray) -> np.ndarray:
+        y = X
+        n_outputs_ = 1 if y.ndim == 1 else y.shape[1]
+        if n_outputs_ != self.n_outputs_:
+            raise ValueError(
+                f"Detected `y` to have {n_outputs_},"
+                f" but this {self} expects"
+                f" {self.n_outputs_} for `y`."
+            )
+        return y
+
+    def inverse_transform(self, X: np.ndarray) -> np.ndarray:
+        y = X  # rename for clarity, the input is always expected to be a target `y`
+        if np.can_cast(self.y_dtype_, np.float32):
```

Comment:
Why are you hardcoding a cast to float32? Why not cast to `y_dtype_`?

Reply:
Sklearn expects regressor outputs to always be float, even if the input target was int. It's checked in one of the sklearn tests I believe.

Reply:
I understand that the output *should* be a float, at least for every case I can think of (I think TF handles complex arrays as two floating point arrays). But I don't think SciKeras should enforce that.

> Sklearn expects regressor outputs to always be float, even if the input target was int. It's checked in one of the sklearn tests I believe.

Do you have a source? I'm not seeing anything type-related in [base.RegressorMixIn][1], [RegressorChain][2], or the definition of [array-like][3].

[1]:https://scikit-learn.org/stable/modules/generated/sklearn.base.RegressorMixin.html#sklearn.base.RegressorMixin
[2]:https://scikit-learn.org/stable/modules/generated/sklearn.multioutput.RegressorChain.html
[3]:https://scikit-learn.org/stable/glossary.html#term-array-like

Reply:
Here's where that is checked: https://github.com/scikit-learn/scikit-learn/blob/73732e5a0bc9b72c7049dc699d69aaedbb70ef0a/sklearn/utils/estimator_checks.py#L2325

I don't think regressors should cast to the input dtype, that's essentially rounding the prediction just because of the input dtype.

Reply:
I'd say that the thing to do here is to _not_ cast at all, but TF returns `float32` arrays and apparently sklearn expects them to be `float64`, at least for multi-output regression: https://github.com/scikit-learn/scikit-learn/blob/73732e5a0bc9b72c7049dc699d69aaedbb70ef0a/sklearn/utils/estimator_checks.py#L1766

Reply:
That's not a test that regressors produce floats from `est.predict(X)`, that's a check to make sure they accept floating point inputs to `est.fit` (or at least it's two lines below).

I see a check for multi-output regressors that the output be float: https://github.com/scikit-learn/scikit-learn/blob/73732e5a0bc9b72c7049dc699d69aaedbb70ef0a/sklearn/utils/estimator_checks.py#L1765-L1767

I'm not seeing a reason to cast the output to float32/64, especially since Keras defaults to float32 IIRC.

Reply:
Yes, TF defaults to float32. We could do:

```python3
if self.y_dtype_ == np.float64 and y.dtype == np.float32:
    return np.squeeze(y.astype(np.float64, copy=False))
return np.squeeze(y)
```

That avoids a cast.

Reply:
There's not much in the way of comments for the sklearn tests, so it's hard to say, but breaking it down, assuming we did cast to the input dtype:
```python3
regressor_1.fit(X, y_)  # y is an int
pred1 = regressor_1.predict(X)  # pred1 is cast to int
regressor_2.fit(X, y_.astype(float))  # y is a float
pred2 = regressor_2.predict(X)  # unless the regressor is perfect, predictions will not be 1.0, 2.0, etc. but rather 1.5, 2.24, etc.
```
So `pred1` != `pred2` and the test fails.

Reply:
Why not avoid a cast all together and let the user decide?

Reply:
> Why not avoid a cast all together and let the user decide?

By that you mean implement https://github.com/adriangb/scikeras/pull/88#discussion_r502992179? We should have at least that one cast to comply with the sklearn expectation for multi-output regressors.

Reply:
> By that you mean implement #88 (comment)?

That's not what I meant, but that looks like the best choice.